Hack week: Plugin update status

0nko opened the pull request:
This is the 2nd part of the plugins hack week project, which adds the plugin version status.

| Light | Dark |
|--------|--------|
| ![image](https://github.com/woocommerce/woocommerce-android/assets/1522856/6e673599-eefa-4c70-959e-6bb766d46fb7) | ![image](https://github.com/woocommerce/woocommerce-android/assets/1522856/94e6fb4d-9960-4813-9c73-12da15bb9cf2) | 

**To test:**
1. Turn on flight mode
2. Go to More menu -> Settings -> Installed plugin
3. Notice an error message with a Retry button appears
4. Turn off flight mode
5. Tap on Retry
6. Notice a list of plugins gets loaded
7. Verify the correct plugin status (Up-to-date, Update available, Inactive) is displayed

0nko committed:
Add a function to fetch installed plugins with update status

0nko committed:
Add string resources

0nko committed:
Update Woo plugin item text color in the settings

0nko committed:
Display the plugin version status in the plugin list

0nko committed:
Fix ktlint issues

0nko committed:
Fix the build

0nko committed:
Add error UI and replace progress bar with shimmer effect

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>PluginStatus</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/plugins/PluginsViewModel.kt:
```diff
@@ -1,44 +1,99 @@
+
+    private fun ImmutablePluginModel.isUpdateAvailable(): Boolean {
+        return when {
+            installedVersion.isNullOrEmpty() || wpOrgPluginVersion.isNullOrEmpty() -> false
```
When using Site credentials, `wpOrgVersion` will be null on the first fetch, because the WPAPI doesn't return this information, and the fetch from wp.org API happens asynchronously ([here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/54e2405183330ae81375b8424cd8ab347f2319d5/fluxc/src/main/java/org/wordpress/android/fluxc/store/PluginStore.java#L905-L907)), with this logic, we wrongly show `up-to-date` for outdated plugins.

WDYT about hiding the update information when `wpOrgPluginVersion` is not available?

0nko replied:
That's a great point. I've created a [PR in FluxC](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2975) that fetches the `wpOrgVersion` information synchronously to avoid this.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/PluginRepository.kt:
```diff
@@ -79,6 +82,28 @@ class PluginRepository @Inject constructor(
+        var loadMore = false
+        do {
+            val payload = FetchPluginDirectoryPayload(SITE, site, loadMore)
+            val action = PluginActionBuilder.newFetchPluginDirectoryAction(payload)
```
This action is not implemented for Jetpack CP sites, and causes a crash, the logic is [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/54e2405183330ae81375b8424cd8ab347f2319d5/fluxc/src/main/java/org/wordpress/android/fluxc/store/PluginStore.java#L960-L969), and as you can see, the WPCom implementation is used only by Jetpack sites, and there is no equivalent for Jetpack CP sites.

I think as a short term solution, we can simply hide the menu for Jetpack CP sites, and later we can work on supporting them using the WPAPI.

0nko replied:
> the logic is [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/54e2405183330ae81375b8424cd8ab347f2319d5/fluxc/src/main/java/org/wordpress/android/fluxc/store/PluginStore.java#L960-L969), and as you can see

As I mentioned in a previous comment, I refactored the way plugins are fetched and I kept the same logic to decide which API to call. I then tested it with a jurassic.ninja site using Jetpack CP and it worked (`!site.isUsingWpComRestApi` [condition](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2975/files#diff-44f37195454337f4b258b698820c5b5fa052a6ee07cc9d21d743e479868b4784R97) applies). 🤔

hichamboushaba replied:
I haven't tested the latest change yet, but I assume that even with the new logic, the app will crash if we sign in using `Jetpack CP`, since the `WPAPI` implementation that you use expects a `username` and a `password` on the `SIteModel` instance, which we won't have when signing using WordPress.com.

But anyway, if you agree on moving to the new `system_status` endpoint, those issues will all be resolved.

hichamboushaba reviewed (changes requested):
Really nice work @0nko, but I found some issues that I shared in the comments.

And I have an additional **non-blocking** concern: for Atomic websites, sometimes they use different versions of some plugins than what's available in wp.org (the plugins I know about are Jetpack and WooCommerce, but it could be more), and these plugins are managed, which means the user can't update them manually. This leads to showing incorrect information about the availability of updates.
For example, below is the state of Jetpack in my current website (WooExpress one):
| App | Web |
| --- | --- |
| <img width="542" alt="Screenshot 2024-03-14 at 14 31 42" src="https://github.com/woocommerce/woocommerce-android/assets/1657201/dafc3136-f7fb-4aa5-b169-9fa21163396f"> | <img width="1188" alt="Screenshot 2024-03-14 at 14 32 54" src="https://github.com/woocommerce/woocommerce-android/assets/1657201/1d4927f0-d647-40b6-8014-0686e9297e5b"> |

I'm not sure what's the best approach for these, as the API doesn't seem to expose something to differentiate them, we can probably hardcode the names of these plugins, and hide the update information for them for Atomic (or WooExpress) sites.

0nko committed:
Remove the async plugin fetching method

0nko committed:
Display the available plugin version in the list

0nko committed:
Fix version comparison

0nko committed:
Update the list of managed plugins

0nko committed:
Don't show the version status for plugins with unknown latest version

0nko commented:
@hichamboushaba Thank you for the thorough review! I've addressed your comments in the code comments. As for the auto-managed plugins, I found a a list of managed plugins (pdKhl6-ds#comment-98) and hardcoded a check. Ready for another round.

Additionally, I found that the version check in `org.wordpress.android.util.helpers.Version` was not working correctly (couldn't handle letters, for example `13.3-a.1), so I copied the algorithm from the web.

0nko committed:
Update plugins preview

0nko committed:
Fix ktlint issues

0nko committed:
Update FluxC reference

hichamboushaba commented:
> As for the auto-managed plugins, I found a a list of managed plugins (pdKhl6-ds#comment-98) and hardcoded a check

Nice find @0nko, but actually I think we can avoid the hardcording, as suggested on the FluxC PR review, the endpoint `/wc/v3/system_status` offers a field `latest_version` that we can use instead of depending on `wp.org` API, and an additional advantage that I just found out about is that the `latest_version` for managed plugins when using this endpoint seems to be matching the Atomic one, so the plugin can correctly show `up-to-date` without any additional logic from our side, for more information check this discussion pdfdoF-4OF-p2#comment-5988

0nko committed:
Update FluxC reference

0nko committed:
Remove auto-managed plugin status and use system plugins for latest version

0nko committed:
Merge branch 'trunk' into hackweek/plugin-status

# Conflicts:
#	build.gradle

0nko commented:
Thank you for looking into this and finding the `system_status` information! 🙇 Like you said, this is much simpler and solves all the problems we faced with the previous approach.

I've update the PR with the FluxC changes and it's ready for a review.

0nko committed:
Unescape plugin names

0nko committed:
Merge branch 'trunk' into hackweek/plugin-status

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/VersionComparison.kt:
```diff
@@ -0,0 +1,50 @@
+package com.woocommerce.android.util
+
+@Suppress("CyclomaticComplexMethod", "LoopWithTooManyJumpStatements", "MagicNumber")
+fun String?.isGreaterThan(v2: String?): Boolean {
```
np, We have a `semverCompareTo` extension [here](https://github.com/woocommerce/woocommerce-android/blob/f3538fb7fed28dcf11916d600a4e52f2139efdd3/WooCommerce/src/main/kotlin/com/woocommerce/android/extensions/StringExt.kt#L58), can't we replace it with this new implementation from you?

But if we are worried that this change wouldn't be safe, I suggest just choosing a more specific name for the extension, `isVersionGreaterThan` should be clearer IMO, WDYT?

0nko replied:
I copied the version comparison from the web implementation of plugins. Looking at the code it seems like it could be specific to the way plugins are versioned and I'm not sure it's applicable in general. It's probably to just rename this function.

hichamboushaba replied:
Unless I'm misunderstanding something from the code, the implementation seems to be a semver comparison, but with an additional logic for additional specifications. And `semverCompareTo` is also used to compare plugin versions, so when using it and not using this new implementation, we are actually missing some scenarios.

This is something we can discuss later, and handle in a separate PR if needed.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/prefs/plugins/PluginsViewModel.kt:
```diff
@@ -43,51 +43,43 @@ class PluginsViewModel @Inject constructor(
+                                Plugin(
+                                    name = StringEscapeUtils.unescapeHtml4(it.name),
+                                    authorName = StringEscapeUtils.unescapeHtml4(it.authorName),
+                                    version = it.version!!,
```
Given what we know about WordPress API quirks, I'm not sure the double bang operator here is always safe 😅

0nko replied:
The `null` versions are filtered out right above it, so it should be fine.

hichamboushaba replied:
Oh, yes, I missed it, sorry.

hichamboushaba commented on WooCommerce/src/main/res/layout/fragment_settings_main.xml:
```diff
@@ -62,7 +62,7 @@
                 <com.google.android.material.textview.MaterialTextView
                     android:id="@+id/woo_plugin_info"
-                    android:textColor="@color/color_on_surface"
+                    android:textColor="@color/color_on_surface_high"
                     android:textAppearance="?attr/textAppearanceBody1"
```
np, I think we should use the appearance `subtitle1` for this without overriding the color, this way it'll match what `WCSettingsOptionValueView` uses.

hichamboushaba reviewed (approved):
Nice work @0nko 🎖️, it works well.
Left some np comments, but feel free to ignore them.

0nko committed:
Change the text appearance of the plugin info item

0nko committed:
Use a more specific version comparison function name

0nko committed:
Update FluxC reference
